### PR TITLE
Do not clear `--platforms` on no-op change to `--cpu`

### PR DESCRIPTION
**Pull request by fmeum**
Resolves a TODO in the code to not unnecessarily fork the configuration when a transition has `--cpu` as a declared output but doesn't change it.

### CODE REVIEW

**Comment by fmeum** [comment]
@gregestren Could you take a look? This resolves a build time regressions we had when we picked up rules_apple's `universal_binary` rule, which performs a no-op transition on `--cpu` on non-Apple platforms.

File: src/main/java/com/google/devtools/build/lib/analysis/starlark/FunctionTransitionUtil.java [hunk 1]
```diff
@@ -127,21 +128,22 @@ static ImmutableMap<String, BuildOptions> applyAndValidate(
...
-    boolean changesCpu = originalOutput.containsKey(COMMAND_LINE_OPTION_PREFIX + "cpu");
-    boolean changesPlatforms = originalOutput.containsKey(COMMAND_LINE_OPTION_PREFIX + "platforms");
-    return changesCpu && !changesPlatforms
-        ? ImmutableMap.<String, Object>builder()
-            .putAll(originalOutput)
-            .put(COMMAND_LINE_OPTION_PREFIX + "platforms", ImmutableList.<Label>of())
-            .build()
-        : originalOutput;
+      BuildOptions options, Map<String, Object> originalOutput) {
+    Object newCpu = originalOutput.get(COMMAND_LINE_OPTION_PREFIX + "cpu");
```
Comment (katre) [inline src/main/java/com/google/devtools/build/lib/analysis/starlark/FunctionTransitionUtil.java:134]: Nit: The names `newCpu` and `originalOutput` don't seem to match. Can we rename one of them?

File: src/test/java/com/google/devtools/build/lib/analysis/StarlarkAttrTransitionProviderTest.java [hunk 2]
```diff
@@ -2501,6 +2501,54 @@ public void testNoPlatformChange() throws Exception {
         .containsExactly(Label.parseAbsoluteUnchecked("//platforms:my_platform"));
   }
 
+  /*
+   * If the transition claims to change --cpu but doesn't, it doesn't constitute a platform change
+   * and also doesn't affect any other options (such as affectedByStarlarkTransition).
+   */
+  @Test
+  public void testCpuNoOpChangeIsFullyNoOp() throws Exception {
```
Comment (katre) [inline src/test/java/com/google/devtools/build/lib/analysis/StarlarkAttrTransitionProviderTest.java:2509]: Thanks for adding a test!

File: src/test/java/com/google/devtools/build/lib/analysis/StarlarkAttrTransitionProviderTest.java [hunk 3]
```diff
@@ -2501,6 +2501,54 @@ public void testNoPlatformChange() throws Exception {
...
+    scratch.file(
+        "platforms/BUILD",
+        "platform(name = 'my_platform',",
+        "    parents = ['" + TestConstants.LOCAL_CONFIG_PLATFORM_PACKAGE_ROOT + ":host'],",
+        "    constraint_values = [],",
+        ")");
+    scratch.file(
+        "test/starlark/my_rule.bzl",
+        "def transition_func(settings, attr):",
+        "  return settings",
```
Comment (katre) [inline src/test/java/com/google/devtools/build/lib/analysis/StarlarkAttrTransitionProviderTest.java:2520]: Add a comment making it clear that CPU isn't changed.

**Review by katre — CHANGES_REQUESTED** [review]
LGTM, just a few small changes I'd like to request.

File: src/main/java/com/google/devtools/build/lib/analysis/starlark/FunctionTransitionUtil.java (hunk 1 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/analysis/starlark/FunctionTransitionUtil.java]: Done, renamed the existing variable to be more descriptive.

**Review by katre — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io flag

**Pull request closed without merging**